tokio-trace-serde/examples/serde_shaved_yak.rs

```diff
@@ -0,0 +1,136 @@
+extern crate serde_json;
+
+#[macro_use]
+extern crate tokio_trace;
+extern crate tokio_trace_core;
+extern crate tokio_trace_serde;
+
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use serde_json::json;
+
+use tokio_trace_core::{
+    event::Event,
+    metadata::{Level, Metadata},
+    span::{Attributes, Id, Record},
+    subscriber::Subscriber,
+};
+
+use tokio_trace_serde::AsSerde;
+
+fn shave(yak: usize) -> bool {
+    span!(Level::TRACE, "shave", yak = yak).enter(|| {
+        debug!(
+            message = "hello! I'm gonna shave a yak.",
+            excitement = "yay!"
+        );
+        if yak == 3 {
+            warn!(target: "yak_events", "could not locate yak!");
+            false
+        } else {
+            trace!(target: "yak_events", "yak shaved successfully");
+            true
+        }
+    })
+}
+
+fn main() {
+    let subscriber = JsonSubscriber {
+        next_id: AtomicUsize::new(1),
+    };
+
+    tokio_trace::subscriber::with_default(subscriber, || {
+        let number_of_yaks = 3;
+        let mut number_shaved = 0;
+        debug!("preparing to shave {} yaks", number_of_yaks);
+
+        span!(Level::TRACE, "shaving_yaks", yaks_to_shave = number_of_yaks).enter(|| {
+            info!("shaving yaks");
+
+            for yak in 1..=number_of_yaks {
+                let shaved = shave(yak);
+                trace!(target: "yak_events", yak = yak, shaved = shaved);
+
+                if !shaved {
+                    error!(message = "failed to shave yak!", yak = yak);
+                } else {
+                    number_shaved += 1;
+                }
+
+                trace!(target: "yak_events", yaks_shaved = number_shaved);
+            }
+        });
+
+        debug!(
+            message = "yak shaving completed.",
+            all_yaks_shaved = number_shaved == number_of_yaks,
+        );
+    });
+}
+
+pub struct JsonSubscriber {
```

Comment:
Take it or leave it: maybe this should appear earlier in the file than the `shave` and `main` functions? Since the example is intended to demonstrate how we would use `tokio-trace-serde` to implement a `Subscriber`, and the `main` and `shave` functions that emit trace events are just to demonstrate that subscriber's output. We provide examples of how to use the `tokio-trace` macros in other crates.

Up to you.

Reply:
That would make the example flow better. I ran into this issue when I was presenting this last night. Did you want me to add other macros in this example as well?

Reply:
> Did you want me to add other macros in this example as well?

What do you mean? I'm fine with re-using the `main` function from the `tokio-trace-fmt` example, the point is just to generate some sample trace events to show how the subscriber works.